java/streams/src/main/java/org/dbsp/circuits/operators/Operator.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (c) 2021 VMware, Inc.
+ * SPDX-License-Identifier: MIT
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package org.dbsp.circuits.operators;
+
+import org.dbsp.circuits.ComputationalElement;
+import org.dbsp.circuits.Wire;
+import org.dbsp.algebraic.dynamicTyping.types.Type;
+import org.dbsp.lib.Linq;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+
+/**
+ * Abstract class representing an operator.
+ * An operator has one or more input wires, and one output wire.
+ * The output wire "belongs" to this operator; the inputs are
+ * references to the outputs of other operators.
+ * An operator also has types for all inputs and for its output.
+ */
+public abstract class Operator extends ComputationalElement {
+    /* final */ Wire output;
+    final List<Wire> inputs;
+    // Number of inputs received in the current computation cycle.
+    int inputsPresent;
+
+    public Operator(List<Type> inputTypes, Type outputType) {
+        super(inputTypes, Linq.list(outputType));
+        this.output = new Wire(outputType, this);
+        // Input wires are not connected.
+        this.inputs = new ArrayList<Wire>(inputTypes.size());
+        for (int i = 0; i < inputTypes.size(); i++)
+            this.inputs.add(null);
+        this.inputsPresent = 0;
+    }
+
+    public void setInput(int index, Wire source) {
+        this.inputs.set(index, source);
+    }
+
+    /**
+     * A new computation cycle has started, reset the internal state.
+     * Most operators do nothing.
+     */
+    public void reset() {}
+
+    public void connectTo(Operator to, int input) {
+        if (to.getInputWires().get(input) != null)
+            throw new RuntimeException(to.getName() + ": input " + input + " already connected");
+        Operator actual = to.getActualConsumer(to, input);
```

Comment:
getActualConsumer should return (Operator, index) pair, where the index is the port number of the actual consumer